src/drivers/esp32/src/esp_slip.c

```diff
@@ -0,0 +1,315 @@
+        }
+        else if (c == 0xDD)
+        {
+          receiverPacket->data[slipDataIndex] = 0xDB;
+          slipDataIndex++;
+        }
+        else
+        {
+          espblReceiveState = error;
+        }
+      }
+      else
+      {
+        receiverPacket->data[slipDataIndex] = c;
+        slipDataIndex++;
+      }
+    }
+    else
+    {
+      if (slipDataIndex == payloadSize)
+      {
+        receiverPacket->status = c;
+      }
+      if (slipDataIndex == payloadSize + 1)
+      {
+        receiverPacket->error = c;
+      }
+      if (slipDataIndex == payloadSize + 3)
+      {
+        espblReceiveState = receiveEnd;
+      }
+      slipDataIndex++;
+    }
+  }
+  break;
+
+  case receiveEnd:
+    if (c == 0xC0)
+    {
+      espblReceiveState = receiveStart;
+      decoderStatus = SLIP_SUCCESS;
+    }
+    else
+    {
+      espblReceiveState = error;
+      decoderStatus = SLIP_ERROR;
+    }
+    break;
+
+  case error:
+    decoderStatus = SLIP_ERROR;
+    break;
+  default:
+    break;
+  }
+  return decoderStatus;
+}
+
+static bool receivePacket(esp_slip_receive_packet *receiverPacket, esp_slip_send_packet *senderPacket, coms_getDataWithTimeout_t getDataWithTimeout, uint32_t timeoutTicks)
+{
+  uint8_t c;
+  uint8_t numberOfUartTimeouts = 0;
+  receiverPacket->status = 1;
+  receiverPacket->command = 0;
+  receiverPacket->dataSize = 0;
+  receiverPacket->value = 0;
+
+  slipDecoderStatus_t packetReceivedStatus = SLIP_DECODING;
+
+  espblReceiveState = receiveStart;
+  while (packetReceivedStatus == SLIP_DECODING && numberOfUartTimeouts < 1)
+  {
+    if (getDataWithTimeout(&c, timeoutTicks))
+    {
+      packetReceivedStatus = decodeSlipPacket(c, receiverPacket, senderPacket);
+    }
+    else
+    {
+      numberOfUartTimeouts += 1;
+    }
+  }
+
+  const uint8_t statusOk = 0;
+  return statusOk == receiverPacket->status && packetReceivedStatus == SLIP_SUCCESS;
+}
+
+static void assembleBuffer(uint8_t *sendBuffer, esp_slip_send_packet *senderPacket)
+{
+  sendSize = senderPacket->dataSize + ESP_OVERHEAD_LEN + 2;
+
+  sendBuffer[0] = 0xC0;
+  sendBuffer[1] = DIR_CMD;
+  sendBuffer[2] = senderPacket->command;
+  sendBuffer[3] = (uint8_t)((senderPacket->dataSize >> 0) & 0x000000FF);
+  sendBuffer[4] = (uint8_t)((senderPacket->dataSize >> 8) & 0x000000FF);
+
+  if (senderPacket->command == FLASH_DATA) // or MEM_DATA
+  {
+    uint32_t checksum = (uint32_t)generateChecksum(sendBuffer, senderPacket);
+    sendBuffer[5] = (uint8_t)((checksum >> 0) & 0x000000FF);
+    sendBuffer[6] = (uint8_t)((checksum >> 8) & 0x000000FF);
+    sendBuffer[7] = (uint8_t)((checksum >> 16) & 0x000000FF);
+    sendBuffer[8] = (uint8_t)((checksum >> 24) & 0x000000FF);
+  }
+  else
+  {
+    sendBuffer[5] = 0x00;
+    sendBuffer[6] = 0x00;
+    sendBuffer[7] = 0x00;
+    sendBuffer[8] = 0x00;
+  }
+
+  sendBuffer[9 + senderPacket->dataSize] = 0xC0;
+}
+
+static void clearUart2Buffer(coms_getDataWithTimeout_t getDataWithTimeout)
```

Comment:
Suggest to rename to avoid the use of "uart2" in the name, maybe flushTxBuffer() or similar. The design with fcn pointers for communication allows for pretty much any mode of communication, not only UART.